source/faucet/main.d

```diff
@@ -382,6 +382,14 @@ public class Faucet : FaucetAPI
         auto owned_utxo_rng = this.state.owned_utxos.byKeyValue()
             // do not pick already used UTXOs
             .filter!(pair => pair.key !in this.used_utxos);
+
+        assert(owned_utxo_rng.walkLength >= 0);
+        if (owned_utxo_rng.walkLength <= 1)
```

Comment:
After this change, we also need to make sure no robots are getting all the UTXOs and starving others from test BOA. Pretty much all Ethereum test faucets have either captcha/facebook post type of authentication.

Reply:
If we do end up in a situation where there will be only 1 UTXO, how will we recover? Do we send 'utxo split' transaction in the current block, so in the next block we will have more than 1 UTXOs?

Reply:
If there is only one `UTXO`, the function throws an `Exception` and does not do anything.
So yes, if the faucet detects this number, it will perform a `split` and there will be > 1 `UTXO`s the next block.

Reply:
you can you popFront() and empty() and avoid using walkLength

Reply:
Do you mean checking if the range is `empty` after a `popFront` call?

Reply:
you only need to know if the owned_utxo_rng array has a length of
0 or
1 or 
more

if you use popFront and empty multiple times in a certain order then you can get the answers for those questions without using walkLength